src/commands/mod.rs

```diff
@@ -1289,13 +1289,15 @@ fn cmd_show(ui: &mut Ui, command: &CommandHelper, args: &ShowArgs) -> Result<(),
         };
     let template_string = format!(
         r#"
-            "Commit ID: " commit_id "\n"
-            "Change ID: " change_id "\n"
-            "Author: " author " (" {author_timestamp_template} ")\n"
-            "Committer: " committer " (" {committer_timestamp_template} ")\n"
-            "\n"
-            if(description, description, {DESCRIPTION_PLACEHOLDER_TEMPLATE} "\n")
-            "\n""#,
+            concat(
+              "Commit ID: {{commit_id}}\n",
+              "Change ID: {{change_id}}\n",
+              "Author: {{author}} ({{{author_timestamp_template}}})\n",
+              "Committer: {{committer}} ({{{committer_timestamp_template}}})\n",
+              "\n",
+              if(description, description, "{{{DESCRIPTION_PLACEHOLDER_TEMPLATE}}}\n"),
+              "\n",
+            )"#,
```

Comment:
I can't decide if I like this better or worse. I don't think I like creating strings and use interpolation inside them, but it does get very verbose without them, assuming we put each part sub-expression on a separate line.

I also don't really like the inconsistency between using string interpolation for each row but using concat() for the overall expression. A single top-level string like this seems more consistent:

```
r#"
"Commit ID: {{commit_id}}
Change ID: {change_id}}
Author: {{author}} ({{{author_timestamp_template}}})
Committer: {{committer}} ({{{committer_timestamp_template}}})

{{if(description, description, "{{{DESCRIPTION_PLACEHOLDER_TEMPLATE}}}
"}}
"
"#
```

But that also has the nested string interpolation and templates I find so hard to read in hg's templates.

I feel like most of the other templates in this commit get harder to read and write. Maybe I'll get used to it, but on the other hand, I never got used to reading hg's templates. What do you think?

Reply:
Suppose templates are written in config file, it's 
* `"Commit ID: {commit_id}\n"` (interpolation), vs 
* `"Commit Id: " ++ commit_id ++ "\n"` (concat operator).

I prefer the former, but maybe you like the latter?

OTOH, if I used `separate("\n", ...)` to concatenate multi-line expressions, I would want to use infix operator to append the terminator: `separate("\n", ...) ++ "\n"`.

> I also don't really like the inconsistency between using string interpolation for each row but using concat() for the overall expression.

Hmm, I didn't consider having multiple ways to concatenate templates is bad idea.

> But that also has the nested string interpolation and templates I find so hard to read in hg's templates.

Yeah, I pretty much hate Mercurial template in that regards.

> I feel like most of the other templates in this commit get harder to read and write.

That's true, but I don't think these templates reflects the real-world use case. They're hard to read partly because we have to use double quotes to quote strings in Rust.

Perhaps, this one is the practical example in which interpolation syntax looks absolutely worse: 
* `-T 'commit_id " " description'` (original)
* `-T '"{commit_id} {description}"'` (string interpolation)
* `-T 'commit_id ++ " " ++ description'` (concat operator)

I might want to use concat operator here.

For this example, I feel both versions look ugly, and I would want to use `concat()` to separate `"   "` from the surrounding expressions.
* `-T 'commit_id.short() "   " description.first_line() " @ " committer.timestamp()'`
* `-T '"{commit_id.short()}   {description.first_line()} @ {committer.timestamp()}"'`

Reply:
I think a concatentation operator would be helpful, especially for something like `separate("\n", ...) ++ "\n"`. Arguably, I think I prefer it to the `concat` function (which could then be replaced with `separate("",...)`). I also like `++` for concatenation, it feels nice an familiar.

Reply:
> I think I'd prefer to also support concatenation without using a function.

Agreed. So the candidates are:
* `++` (borrowed from haskell)
* `~` (borrowed from pest)

I'm pretty sure `+` is bad here because of ambiguity between `int+int` and `whatever+str`.

> switch to making the top-level expression a string like hg does

I think it's another wart of hg template. `-Tidentifier` could be parsed as `"identifier"` or named template `identifier`.

> even support a trailing concatenation operator to make it easier to use in multi-line templates, or do you think that would cause any issues with parsing?

I think pest can deal with that, but I prefer not adding noop postfix operator that looks like an infix op.